lib/RGBot.js

```diff
@@ -400,39 +769,142 @@ const RGBot = class {
+     * and attack the target 1 time.  To finish off a target, this method must be called until the target is dead.
+     *
+     * @param {Entity} entity - The entity to attack
+     * @param {object} [options={}] Optional parameters
+     * @param {number} [options.reach=2] How close to get to the target before attacking
+     * @param {Item} [options.attackItem=undefined] An item in the bot inventory to use for the attack
+     * @returns {boolean} - did we successfully attack
+     *
+     * @example
+     * let target = //<someEntity>
+     * while (target.isValid) {
+     *     attackEntity(target)
+     * }
+     *
+     * Note: This method currently only uses melee weapons
      */
-    attackEntity(entity) {
-        if (!entity) {
-            console.error(`attackEntity: Entity was null or undefined`);
-        } else {
+    async attackEntity(entity, options={}) {
+        const reach = options.reach || 2;
+        if (entity) {
             try {
-                this.#log(`Attacking ${this.getEntityName(entity)}`);
-                this.#bot.attack(entity, true);
-            } catch (err) {
-                console.error(`Error attacking target: ${this.getEntityName(entity)}`, err)
+                this.#logDebug(`Following and attacking entity ${entity.username || entity.name}`)
+                if (entity.isValid) {
+                    const moveGoal = new GoalNear(entity.position.x, entity.position.y, entity.position.z, reach)
+                    try {
+                        // if this fails, then it throws an exception
+                        await this.#bot.pathfinder.goto(moveGoal)
+                        await this.followEntity(entity, {reach: reach})
+                        await this.waitForWeaponCoolDown()
+                        // TODO: Verify if the target can be attacked melee (flying creatures can't be)
+                        let attackItem = options.attackItem || this.bestAttackItemMelee()
+                        if (attackItem) {
+                            // TODO: Check enchants or exact instance id of the item instead of name
+                            if (!this.#bot.heldItem || attackItem.name !== this.#bot.heldItem.name) {
+                                this.#log(`Equipping attack weapon: ${attackItem.name}`)
+                                await this.#bot.equip(attackItem, 'hand');
+                            }
+                        }
+                        this.#logDebug(`Attacking entity: ${entity.username || entity.name} using weapon: ${attackItem?.name}`)
+                        this.lastAttackTime = Date.now()
+                        this.lastAttackItem = attackItem
+                        this.#bot.attack(entity)
+                        return true
+                    } catch (ex) {
+                        this.#logDebug(`Failed to get near enough to entity to attack`);
```

Comment:
Is this always the case? Can it fail before if the player is unable to equip an item, for instance?